Code/Framework/AzCore/AzCore/Jobs/JobExecutor.cpp

```diff
@@ -0,0 +1,361 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/Jobs/JobExecutor.h>
+#include <AzCore/Jobs/JobGraph.h>
+
+#include <AzCore/std/string/string.h>
+#include <AzCore/std/containers/queue.h>
+#include <AzCore/std/parallel/binary_semaphore.h>
+#include <AzCore/std/parallel/semaphore.h>
+#include <AzCore/std/parallel/mutex.h>
+#include <AzCore/std/parallel/scoped_lock.h>
+#include <AzCore/std/parallel/thread.h>
+
+#include <random>
+
+namespace AZ
+{
+    constexpr static size_t PRIORITY_COUNT = static_cast<size_t>(JobPriority::PRIORITY_COUNT);
+
+    namespace Internal
+    {
+        CompiledJobGraph::CompiledJobGraph(
+            AZStd::vector<TypeErasedJob>&& jobs,
+            AZStd::unordered_map<uint32_t, AZStd::vector<uint32_t>>& links,
+            size_t linkCount,
+            bool retained)
+            : m_remaining{ jobs.size() }
+            , m_retained{ retained }
+        {
+            m_jobs = AZStd::move(jobs);
+            m_dependencyCounts = reinterpret_cast<AZStd::atomic<uint32_t>*>(azcalloc(sizeof(AZStd::atomic<uint32_t>) * m_jobs.size()));
```

Comment:
any reason you couldnt use a vector for the atomics? there is a `reserve_no_construct`

Reply:
So, I might have been overzealous in trying my best not to add things to the header I didn't need. I'll revisit this most likely